index.js

```diff
@@ -48,16 +48,45 @@ function ensureAllCommentsPrinted(astComments) {
   });
 }
 
-function format(text, opts) {
+function format(text, opts, addIndents) {
+  addIndents = addIndents || 0;
+
+  const rangeStart = opts.rangeStart;
+  const rangeEnd = Math.min(opts.rangeEnd, text.length)
+
+  if (0 < rangeStart || rangeEnd < text.length) {
+    const rangeString = text.substring(rangeStart, rangeEnd)
+    const numIndents = countIndents(rangeString, opts);
+
+    const rangeFormatted = format(rangeString, Object.assign({}, opts, {
+      rangeStart: 0,
+      rangeEnd: Infinity,
+      printWidth: opts.printWidth - numIndents * opts.tabWidth
+    }), numIndents);
+
+    return text.slice(0, rangeStart) + rangeFormatted + text.slice(rangeEnd)
+  }
+
   const ast = parser.parse(text, opts);
   const astComments = attachComments(text, ast, opts);
-  const doc = printAstToDoc(ast, opts);
+  const doc = printAstToDoc(ast, opts, addIndents)
   opts.newLine = guessLineEnding(text);
   const str = printDocToString(doc, opts);
   ensureAllCommentsPrinted(astComments);
+  // Remove extra leading newline as well as the added indentation after last newline
+  if (addIndents > 0) {
+    return str.slice(opts.newLine.length).trimRight() + opts.newLine;
+  }
   return str;
 }
 
+function countIndents(text, opts) {
```

Comment:
If you look at template literal code, I have some piece of code that is much more precise.